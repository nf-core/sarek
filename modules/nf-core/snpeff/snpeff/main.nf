process SNPEFF_SNPEFF {
    tag "$meta.id"
    label 'process_medium'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/snpeff:5.1--hdfd78af_2' :
        'biocontainers/snpeff:5.1--hdfd78af_2' }"

    input:
    tuple val(meta), path(vcf)
    val   db
    tuple val(meta2), path(cache)

    output:
    tuple val(meta), path("*.ann.vcf"),   emit: vcf
    tuple val(meta), path("*.csv"),       emit: report
    tuple val(meta), path("*.html"),      emit: summary_html
    tuple val(meta), path("*.genes.txt"), emit: genes_txt
    path "versions.yml"                 , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def avail_mem = 6144
    if (!task.memory) {
        log.info '[snpEff] Available memory not known - defaulting to 6GB. Specify process memory requirements to change this.'
    } else {
        avail_mem = (task.memory.mega*0.8).intValue()
    }
    def prefix = task.ext.prefix ?: "${meta.id}"
    def cache_command = cache ? "-dataDir \${PWD}/${cache}" : ""
    """
    snpEff \\
        -Xmx${avail_mem}M \\
        $db \\
        $args \\
        -csvStats ${prefix}.csv \\
        $cache_command \\
        $vcf \\
        > ${prefix}.ann.vcf

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        snpeff: \$(echo \$(snpEff -version 2>&1) | cut -f 2 -d ' ')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}.ann.vcf

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        snpeff: \$(echo \$(snpEff -version 2>&1) | cut -f 2 -d ' ')
    END_VERSIONS
    """

}
