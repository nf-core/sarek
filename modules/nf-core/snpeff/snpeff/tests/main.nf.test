nextflow_process {

    name "Test Process SNPEFF_SNPEFF"
    script "../main.nf"
    process "SNPEFF_SNPEFF"
    config "./nextflow.config"
    tag "modules"
    tag "modules_nfcore"
    tag "modules_snpeff"
    tag "snpeff"
    tag "snpeff/download"
    tag "snpeff/snpeff"

    test("test_SNPEFF_SNPEFF") {

        setup {
            run("SNPEFF_DOWNLOAD") {
                script "../../download/main.nf"
                process {
                    """
                    input[0] = Channel.of([[id:params.snpeff_db], params.snpeff_db])
                    """
                }
            }
        }

        when {
            process {
                """
                input[0] = Channel.of([
                    [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/vcf/test.vcf', checkIfExists: true)
                ])
                input[1] = params.snpeff_db
                input[2] = SNPEFF_DOWNLOAD.out.cache
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert path(process.out.report[0][1]).exists() },
                { assert path(process.out.summary_html[0][1]).exists() },
                { assert path(process.out.vcf[0][1]).exists() },
                { assert snapshot(process.out.genes_txt).match("genes_txt") },
                { assert snapshot(process.out.versions).match("versions") }
            )
        }
    }
}
