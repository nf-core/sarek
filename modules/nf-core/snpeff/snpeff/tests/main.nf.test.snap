{
    "versions": {
        "content": [
            [
                "versions.yml:md5,25d44a118d558b331d51ec00be0d997c"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.02.0"
        },
        "timestamp": "2024-03-18T17:37:18.879477"
    },
    "genes_txt": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "test.genes.txt:md5,130536bf0237d7f3f746d32aaa32840a"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.02.0"
        },
        "timestamp": "2024-03-18T17:37:18.874822"
    }
}